### PR TITLE
HPCC-12990 Allow external c++ files to be compiled into queries

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday gavin.halliday@lexisnexis.com

### CODE REVIEW

**Comment by ghalliday** [comment]
@richardkchapman @jamienoss for discussion.  This allows c and c++ files to be included in compiles, but doesn't solve the problem of including the files in an archive.

File: ecl/hqlcpp/hqlcpp.cpp [hunk 1]
```diff
@@ -1226,6 +1234,8 @@ bool HqlCppInstance::useFunction(IHqlExpression * func)
     }
     if (include.length())
         useInclude(include.str());
+    if (source.length())
```
Comment (jamienoss) [inline ecl/hqlcpp/hqlcpp.cpp]: This makes me wonder, would it be an idea to add isEmpty() { return !buffer||!*buffer; } to StringBuffer as there is for StrngAttr? I will open a jira as a suggestion.

Comment (jamienoss) [inline ecl/hqlcpp/hqlcpp.cpp]: https://track.hpccsystems.com/browse/HPCC-13077

File: ecl/hqlcpp/hqlecl.cpp [hunk 2]
```diff
@@ -490,11 +489,21 @@ void HqlDllGenerator::doExpand(HqlCppTranslator & translator)
             expandCode(CHILD_MODULE_TEMPLATE, fullext, code, true, i+1, translator.queryOptions().targetCompiler);
 
             StringBuffer fullname;
-            fullname.append(wuname).append("_").append(i+1);
+            fullname.append(wuname).append("_").append(i+1).append(".cpp");
             sourceFiles.append(fullname);
         }
     }
 
+    for (unsigned iExtra=0; ; iExtra++)
```
Comment (jamienoss) [inline ecl/hqlcpp/hqlecl.cpp]: Superficial, but is this a style change rather than using loop?

File: system/jlib/jcomp.cpp [hunk 3]
```diff
@@ -483,11 +484,11 @@ bool CppCompiler::compileFile(IThreadPool * pool, const char * filename, Semapho
     {
         if (targetDir.get())
             cmdline.append(" /Fo").append("\"").append(targetDir).append("\"");
-        cmdline.append(" /Fd").append("\"").append(targetDir).append(createDLL ? SharedObjectPrefix : NULL).append(filename).append(".pdb").append("\"");//MORE: prefer create a single pdb file using coreName
+        cmdline.append(" /Fd").append("\"").append(targetDir).append(createDLL ? SharedObjectPrefix : NULL).append(basename).append(".pdb").append("\"");//MORE: prefer create a single pdb file using coreName
```
Comment (jamienoss) [inline system/jlib/jcomp.cpp]: Does this comment still apply?

**Comment by jamienoss** [comment]
@ghalliday & @richardkchapman What's here looks ok to me, though I'm really not sure on how complete it is? Obviously needs a test if intended to fly without archive functionality.

**Comment by ghalliday** [comment]
@richardkchapman rebased on top of your changes - which removed most of the significant differences.

**Comment by richardkchapman** [comment]
What's the plan about archive handling?

**Pull request merged**